src/python/pants/backend/python/util_rules/local_dists_pep660.py

```diff
@@ -0,0 +1,539 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import os
+import shlex
+from collections import defaultdict
+from dataclasses import dataclass
+
+# TODO: move this to a util_rules file
+from pants.backend.python.goals import setup_py
+from pants.backend.python.goals.setup_py import (
+    DependencyOwner,
+    ExportedTarget,
+    OwnedDependencies,
+    create_dist_build_request,
+)
+from pants.backend.python.subsystems.setup import PythonSetup
+from pants.backend.python.subsystems.setuptools import PythonDistributionFieldSet
+from pants.backend.python.target_types import PythonProvidesField, PythonResolveField
+from pants.backend.python.util_rules.dists import (
+    BuildBackendError,
+    DistBuildRequest,
+    distutils_repr,
+)
+from pants.backend.python.util_rules.dists import rules as dists_rules
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.base.build_root import BuildRoot
+from pants.core.util_rules import system_binaries
+from pants.core.util_rules.system_binaries import BashBinary, UnzipBinary
+from pants.engine.fs import (
+    CreateDigest,
+    Digest,
+    DigestSubset,
+    FileContent,
+    MergeDigests,
+    PathGlobs,
+    RemovePrefix,
+    Snapshot,
+)
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import AllTargets, Target, Targets, WrappedTarget, WrappedTargetRequest
+from pants.engine.unions import UnionMembership
+from pants.util.docutil import doc_url
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.osutil import is_macos_big_sur
+from pants.util.strutil import softwrap
+
+logger = logging.getLogger(__name__)
+
+
+# We just use DistBuildRequest directly instead of adding a PEP660BuildRequest
+
+
+@dataclass(frozen=True)
+class PEP660BuildResult:  # based on DistBuildResult
+    output: Digest
+    # Relpaths in the output digest.
+    editable_wheel_path: str | None
+
+
+# PEP 660 defines `prepare_metadata_for_build_editable`. If PEP 660 is not
```

Comment:
If PEP 660 is not supported by whom? This is for in-repo, first-party code, right?  So we could require that you support PEP 660 (and maybe the PEP 517 fallback)? It is in the user's power to make that so. 

I'm not necessarily proposing changes, more trying to understand the context, and how much bad stuff in the wild this is expected to handle?

Reply:
PEP 660 is fairly new - it was marked as final only 10 months ago, which is a fairly short period of time for python packaging.

I'm still using python3.6 (sadly), and the version of setuptools that supports python3.6 does not support PEP 660. And, that older version of setuptools generated dist-info directories named `pkg_name.dist-info` instead of `pkg_name-version.dist-info`. So, I needed to handle most of these quirks for st2 where I want to use this feature.

The one quirk I didn't encounter was falling back to `build_wheel`. I have not reviewed the implementations of anything other than setuptools, so if there is a version (prob an older version) of poetry, pdm, flint, or whatever that people have configured as their PEP-517 backend, then this is required to fully support the PEP-517 and PEP-660 specs. In [this section](https://peps.python.org/pep-0517/#appendix-a-comparison-to-pep-516) of PEP 517, it highlights:

> One specific consequence of this is that in this PEP, we’re able to make the `prepare_metadata_for_build_wheel` command optional. In our design, this can be readily handled by build frontends, which can put code in their subprocess runner like:

And then it shows a code block that falls back to `build_wheel` and then runs an undefined `unzip_metadata` function to get the metadata.

> So we could require that you support PEP 660 (and maybe the PEP 517 fallback)? It is in the user's power to make that so.

The backend does not actually have to support PEP 660 because we cannot delegate building the wheel to an external backend. We cannot delegate building the editable wheel because we're running in a sandbox and the spec doesn't give us a way to provide which directory should be used as the editable sources. So, all we're using the backend for is to get the wheel metadata. Though it does not need to implement PEP 660, it must support PEP 517. Basically, Pants becomes the PEP 660-compliant frontend + backend, delegating metadata creation to the user-requested PEP-517 backend.